### ISSUE TITLE
Phase 3: Collaboration > Media Library

### ISSUE THREAD

**Issue by ramonjd** [WordPress/gutenberg#55238]
Tasks that are a part of the Phase 3 [Media Library work described here](https://make.wordpress.org/core/2023/07/07/media-library/).

## Design

![Media](https://github.com/WordPress/gutenberg/assets/846565/09595fe6-d0aa-4e73-8b5d-8ce3ea376d7f)

### i1 — parity for the existing media library

- Use [DataViews](https://github.com/WordPress/gutenberg/issues/55083) to implement a Media screen. With fields including:
  - Media preview
  - Author
  - Upload date
  - File name
  - File size
  - Pixel dimensions
  - Media type (image, audio, pdf, etc.)
  - Attached to
- Uploading via _Add media_ button and drag and drop.
- A Media details / `MediaEditor` view, using DataView Forms. Allow updating metadata including:
  - Title
  - Caption
  - Alt text
  - Description
- Within the Media details / `MediaEditor` view, implement an image cropper that supports:
  - Aspect ratio
  - Crop
  - Rotation
  - Flip
  - Tracking edits associated with the `attachment` entity (for eventual support of undo/redo, and resetting state)
    - To support this, we might need the core data store to support clearing edits directly?
- Within the block editor, add a `MediaPicker` that implements a picker version of DataViews for selecting one or many media items.
  - Used for selecting images in the Image block, setting a Featured Image, setting a background image, or selecting media for the Audio, Video, and Gallery blocks
  - Support filtering by `Attached to` to show all media that is attached to the current post/page

### i2 — beyond 

- Categories and tags support. Some investigation took place in #53788.

Completed

- [List View: Add media previews to list view for gallery and image blocks #53381](https://github.com/WordPress/gutenberg/pull/53381)

## Related reading

- [Media Library - Make WordPress Core post](https://make.wordpress.org/core/2023/07/07/media-library/)
- #53788
- #59745
- #61447
- #58172
- https://github.com/WordPress/gutenberg/issues/66663
- https://github.com/WordPress/gutenberg/issues/62371

**Comment by antpb** [comment]
Hello! in a recent Media component meeting we decided to start flagging trac tickets with `phase-3-media-triage` so we can in a future meeting discuss the impact of the Media Library revamp relative to the flagged issues. We see a lot of opportunity in this initiative to solve longstanding Media issues that have been held back by either backbone of the overall data structure of media items. 

One that I was just reviewing for instance: [42751](https://core.trac.wordpress.org/ticket/42751) gives us a really good indicator of plugins that are extending the current media library frames. We could check for anyone deregistering the mediaelement css. 

More to come as we parse through trac. Just wanted to share our efforts to help align our component focus with this initiative. Please feel free to join any future Media component meetings ( Every Wednesday at 15:00 UTC ) we'd be very open to changing or alternating the time if the current one isn't good for anyone working on the project.

**Comment by David-Lazo** [comment]
Allow us to organize media in Folders which we can also color code and bake the functionality of these two plugins into WordPress Core.
https://wordpress.org/plugins/enable-media-replace/
https://wordpress.org/plugins/post-types-order/

Also add an option to download all media.

Competing products:
Happy Files:https://happyfiles.io/

FileBird:https://codecanyon.net/item/media-folders-manager-for-wordpress/21715379

Take inspiration from Eagle Application for some great media filtering ideas: https://en.eagle.cool/

**Comment by shoelaced** [comment]
Would love to have the ability to control responsive image cropping. Here's a great example of why:
https://youtu.be/fp9eVtkQ4EA?si=3ApM7CS569t0iQqB&t=614

Basically: Please bring back the thumbnail editor and further allow isolated edits to all of the other image sizes.

**Comment by jameskoster** [comment]
I updated the designs in the OP based on the latest data view work. We should decide:

* Default views (e.g. 'All media', 'Images', 'Video', etc.)
* Default layout (grid or list).
* Primary actions.
* Fields to include (mockups are based on wp-admin, but it might be nice to include things like file size/type).
* Which fields should be visible by default.
* What filters users can enact.
* Default setting for 'per page' view option.

**Comment by noisysocks** [comment]
Sat down with @andrewserong, @ramonjd, and @tellthemachines and had a go at scoping this into actionable tasks. I've updated the description.

@jameskoster – do you have a design for the details view?

**Comment by youknowriad** [comment]
> Use https://github.com/WordPress/gutenberg/issues/55083 to implement Media screen in site editor.

@noisysocks I think we shouldn't be overloading the site editor with media screen (same for posts), we want to instead try to replace to "Media" menu item top level in WP-Admin with a site editor like experience. 

I know it's harder but if we add media, posts... to the site editor, the site editor won't make sense anymore under appearance and we should ideally keep its scope about "design/appearance".

Granted that "pages" is misplaced with this logic but I believe it should ideally move as well and both "Media" and "Posts" constitute good opportunities to start exploring this.

**Comment by noisysocks** [comment]
@youknowriad Yeah true, this is probably an opportune time to start moving things out.

**Comment by noisysocks** [comment]
Thinking out loud. The problem with replacing WP Admin → Media is that it means we need a level of feature parity before making the switch. I'd like to ship this functionality iteratively if possible so that we're getting continuous feedback. 

Maybe an option is to have the _Media library_ button open the new DataViews media library in a modal, or something. That way it gets testing and we can build confidence in it.

<img width="730" alt="Screenshot 2024-05-17 at 14 22 20" src="https://github.com/WordPress/gutenberg/assets/612155/998afd57-201d-4cdc-bb38-b6a7a4da6c2e">

**Comment by youknowriad** [comment]
I think we should just start as an experiment to enable, and once we reach close to feature parity, remove the flag.

**Comment by fabiankaegy** [comment]
I know priority wise this is going to be one of the last pieces. But a very common thing we do today in the media library / media library modal is extend the available field for a given piece of media to store additional image metadata to post meta. Things like image credit come up very often etc.

**Comment by noisysocks** [comment]
Yes good point we'll need to support custom fields. Are they added via `attachment_fields_to_edit`? I'm not sure how it works. **edit:** yep, [here's](https://code.tutsplus.com/creating-custom-fields-for-attachments-in-wordpress--net-13076a) a good reference. Fortunately it looks a lot more structured than post custom fields which were a nightmare to support in the block editor...

**Comment by jameskoster** [comment]
> do you have a design for the details view?

Do you mean the Inspector panel? If we follow the principles outlined in https://github.com/WordPress/gutenberg/issues/59689 then we end up with something like this:

<img width="1314" alt="media details" src="https://github.com/WordPress/gutenberg/assets/846565/8230690d-6a87-4f38-8f1d-a49901379bc4">

* 'Categories' and 'Tags' are new, and can probably be omitted initially.
* 'Caption' and 'Alt text' can open dialogs similar to editing a post excerpt.
* 'Comments' should link to a filtered view of the Comments table.
* 'Attached to' could open a modal containing something like a searchable listbox. We might reuse the UI you see when creating a template for a specific entity.
* Template panel would be the same as posts pages. This isn't built yet. The current interface would be okay in the interim.

**Comment by youknowriad** [comment]
Another thing I wanted to mention here, is that it's important to stop creating custom components for all these "inspectors" and instead start exploring a more systemic approach like outlined here  https://github.com/WordPress/gutenberg/issues/59745

**Comment by noisysocks** [comment]
> Do you mean the Inspector panel? If we follow the principles outlined in #59689 then we end up with something like this:

Sorry, I meant the page that appears when you click on an image. It looks like this in the current media library:

<img width="1624" alt="Screenshot 2024-05-20 at 09 31 40" src="https://github.com/WordPress/gutenberg/assets/612155/838290cc-3fd3-4c11-a1dc-4d60053b1807">

**Comment by jameskoster** [comment]
Ah yes. I don't think this is 100% concrete so I would welcome feedback, but my expectation is that you'd be taken to an Editor instance with image editing tools, something like: 

<img width="1595" alt="Image editor" src="https://github.com/WordPress/gutenberg/assets/846565/012f69fd-c743-4e60-b83e-4fe9cfd4d702">

The actual flows for cropping and scaling would need some design exploration. 

Other media; audio, archives, sheets, probably may not need a dedicated editor with these tools. Such files would be modified via Inspector:

<img width="1742" alt="audio" src="https://github.com/WordPress/gutenberg/assets/846565/b07ca7f6-3629-48dc-94e6-eba56427a0c6">

**Comment by mrkdevelopment** [comment]
I think splitting the side bar a bit would make sense.

1. Title, Descriptions, alt and caption are all text base information. These could be one popover for editing (and would be easy in that cause to add to for AI description generation).

2. Attached, comments, category and tag are related fields. Again, could be one popover for all values.

3. File fields (size, format, dimensions, filename) these are all generally not edited.

The reason I like a single popover is it will save clicks. When you edit a title, alt text and description - opening and closing multiple popovers would be a pain and takes a lot more time.

Ideally the side bar would be big enough to have tabs and you would not even need to open a popover.

**Comment by noisysocks** [comment]
Thanks @jameskoster. Should be enough to start getting the rough shape of this built behind an experimental flag 👍

**Comment by mtias** [comment]
I really think the details panel should be a separate tile, like the side by side view we have for the preview. This is also what we had been discussing for "quick edits" replacement.

**Comment by jameskoster** [comment]
A separate tile makes sense. Theoretically this enables the Details panel to be invoked from the side-by-side view as well. 

<img width="1952" alt="Media" src="https://github.com/WordPress/gutenberg/assets/846565/5de98f51-4f1d-4098-867b-8c2910b217ff">

It could also be worth exploring greater structural alignment between data views / full-screen editor.

**Comment by rmorse** [comment]
Is there a discussion somewhere where it was decided that all the new UI coming to the project (which looks great btw) would be dropping support for a persistent wp-admin sidebar?

I find it frustrating to have to click that back arrow, then again on the WP/site icon, and wait for the page reload, just to load the admin sidebar UI so I can navigate around.

**Comment by jameskoster** [comment]
@rmorse you may be interested in [this post](https://make.wordpress.org/core/2023/07/12/admin-design/), which touches on navigation (and personalisation thereof). I agree that a critical detail to get right in the design is facilitating quick-access to the top-level of navigation.

**Comment by james-tyner** [comment]
Hello @jameskoster! I appreciate the level of care you're putting into this process.

My primary experience with CMSes is at large news publishers, and I've heard complaints from many photo editors over time. I have some thoughts/questions re: how to ensure the Media Library experience is fast and intuitive for people who spend all day in it.

**First:** Are there plans to support both a list view and a grid/thumbnail view for assets in the Media Library? Looking at the mockups above, I'm not sure where I would find the toggle to switch between those. (To be fair, I'm not well-versed in the new/upcoming WordPress design system…)

The two views would support different types of users. In a large newsroom, users who are trying to identify the best image to add to an article would likely seek it out using the thumbnail view, while photo editors who are working with photo metadata might prefer the list view if it reveals more metadata without clicking. So it would be good for users to be able to choose one of the two views and have that choice persist over time.

**Second:** I agree with @mrkdevelopment that minimizing the number of clicks required to edit the key metadata is really important, especially for people who spend all day working with those fields. I'd like to add that it would also be valuable for those fields to be more readily scannable, and it seems like the current designs only make visible, at most, a few dozen characters before clicking.

<img width="317" alt="image" src="https://github.com/WordPress/gutenberg/assets/45050998/55017c53-b68f-40b0-802a-c0f4696dd5fb">

I wonder if the visible text can be expanded a bit more without disrupting the layout, and I'm curious to know what we could expect the UI to look like once a user clicks on one of those fields. I'd especially advocate for the entire caption, description, or alt text to be made visible at once upon clicking, rather than requiring additional scrolling to read — I know the scrolling and/or dragging required to read longer captions in the currently live Media Library design is frustrating for photo editors.

Captions and descriptions are sometimes written to include additional information for internal purposes, so it's important that the entire text of those fields is easy to access.

**Third:** As @fabiankaegy noted, it's common to add custom fields to these media items. If we end up grouping fields together in different media views, it would be valuable for developers to be able to specify where those custom fields should appear/not appear so they're grouped and ordered logically with each other and with the core fields. (This may be something that's easy to do already that I'm not familiar with)

Thank you for all your work on this 🙏

**Comment by jameskoster** [comment]
> Are there plans to support both a list view and a grid/thumbnail view for assets in the Media Library?

Absolutely. To your point about where to find those controls, a change was recently merged to make this more obvious: https://github.com/WordPress/gutenberg/pull/63205

> I wonder if the visible text can be expanded a bit more without disrupting the layout

_For now_ the plan is to follow the design patterns as implemented in the Inspector component. So to view a value you'd have to click it to open the associated popover. These details can change though, and may need to as this UI proliferates out of the full-screen editor context.

**Comment by davidperezgar** [comment]
Thank you for this project! I can see how necessary it is.
Working with native image sizes is important for a variety of reasons. I'd like to see:
- an interface where you can add new image sizes;
- The ability to select the image size when using an image block.
- The option to define cropped images as image sizes.

Are all of these features planned to address the current issues?

<img width="1146" height="753" alt="Image" src="https://github.com/user-attachments/assets/c7734666-3b3d-4f4a-bba8-689c587fbc69" />

**Comment by andrewserong** [comment]
Thanks for laying that out @davidperezgar — no, I don't believe there's an issue or plan for a UI to manage custom image sizes at this stage, though that _is_ a cool idea. Feel free to open a separate issue for it if you'd like! (For now, just having the comment here is a good way to flag it, though 👍)

For my part, where I'm chipping in right now is with both the experimental media modal and media editor (I've assigned myself to a couple of sub-issues from this tracking issue). My goal is to chip away at building out feature parity with the existing screens in core, and once we've got feature parity, then start looking into which cool new features we could add in.

**Comment by jeffpaul** [comment]
@andrewserong I'd like to explore integrating the [AI Experiments plugin](https://github.com/wordPress/ai) with the experimental media modal and media editor; if there's anything from the plugin's current featureset or things we might plan in the future that you think would make a good case to integrate with your work, then I'd love to hear your thoughts... thanks!

**Comment by andrewserong** [comment]
Thanks for the ping @jeffpaul! This is a good time to start considering how AI experiments might extend these interfaces (and where and how the interfaces will need to be adjusted in order to support those points of extension). A few thoughts:

* If the extension is about specific fields to be injected or replaced with AI-powered alternatives then something to follow will be the work @oandregal has started looking into for https://github.com/WordPress/gutenberg/issues/74865 — i.e. if we're registering fields for entities on the server, then maybe there'll be a server-side way to override or inject the AI-powered fields you might need
* If the extension is about injecting buttons or other things into the UI itself, then this raises some interesting questions about how we want to go about it. E.g. whether that's via slots or other mechanisms. There's an interesting discussion about that kind of thing in the content guidelines work, too: https://github.com/WordPress/gutenberg/pull/76338
* For the media editor in particular, things are much earlier along than with the media modal. I have an older PR that I'm planning to dust off once 7.0 is done, to look at hooking up cropping tools: https://github.com/WordPress/gutenberg/pull/74796 — that PR isn't ready for review yet, but hopefully it gives an idea of the direction I had in mind. @ramonjd has been exploring some ideas in that area, too, and something that would be cool is for editing / cropping operations to have a standard format so that AI tools could perform actions without having to use the UI

Did you have ideas already for how you'd like to integrate the AI experiments with these experiments? If so, it could be good to open a separate issue (we can link this issue as the parent) so we can discuss what might need to happen in the modal and media editor in order to support it.

One thing to keep in mind is that we're still heavily changing the UI and structure of the modal and media editor while it's behind an experiment, so it'll be good to factor these sorts of things in early(ish) to make sure we don't accidentally break anything as we progress.

For the media editor, I still have quite a few questions about _how_ it integrates with the block editor that I'd like to explore (as part of the WIP PR in #74796), so at this stage I possibly have more questions than answers for some of this stuff 😄

**Comment by jeffpaul** [comment]
> Did you have ideas already for how you'd like to integrate the AI experiments with these experiments?

Most immediately it would likely be looking to integrate existing AI Experiments plugin functionality that already touches on images (i.e. [Image](https://github.com/WordPress/ai/releases/tag/0.3.0) [Generation](https://github.com/WordPress/ai/releases/tag/0.4.0), [Alt Text Generation](https://github.com/WordPress/ai/releases/tag/0.3.0)) and then the additional features that are planned (e.g. [Image Editing](https://github.com/WordPress/ai/issues/236), [Background Removal](https://github.com/WordPress/ai/issues/237), [Smart Cropping](https://github.com/WordPress/ai/issues/238), [Image Tagging](https://github.com/WordPress/ai/issues/45)).  Sounds like it may be best for us to sit back and await the media editor and modals to be more stable before trying to inject there as things are still moving around?

**Comment by andrewserong** [comment]
> Sounds like it may be best for us to sit back and await the media editor and modals to be more stable before trying to inject there as things are still moving around?

Quite possibly!

For things like alt text generation (or captions, description), I suspect the place to hook in will be when the entity-based field registration work progresses.

For image generation, editing, background removal, cropping, and tagging, yes, I think we'll want to see how the shape of the editor looks before figuring out exactly where and how it can be extended. That said, if you have any ideas for how the integration points should ideally work, please share them here (or in any other issues) as that'll help inform how we go about it.

All that said, thanks very much for sharing the issues! That already helps flag that in the media editor we'll need clear places to inject those sorts of controls and actions 👍
